### PR TITLE
CMake: Fallback from try_run to try_compile when cross-compiling.

### LINKED ISSUE #432: Make cross-compiling easier
[cmake/CXXFeatureCheck.cmake](https://github.com/google/benchmark/blob/master/cmake/CXXFeatureCheck.cmake) uses try_run, which is bound to fail when cross-compiling. The [try_run documentation](https://cmake.org/cmake/help/git-master/command/try_run.html) warns that the command will be a pain in cross-compiling, and suggests guarding it with an `if(CMAKE_CROSSCOMPILING)` block. Will you please consider adopting that suggestion?

Background: I'm including this in a project of mine, and I'm trying to cross-compile for Android using `-DCMAKE_SYSTEM_NAME=Android -DCMAKE_ANDROID_NDK=$HOME/Library/Android/sdk/ndk-bundle -DCMAKE_ANDROID_ARCH_ABI=arm64-v8a -DCMAKE_BUILD_TYPE=Release`. I was able work this issue by adding `-DRUN_HAVE_POSIX_REGEX=0`, but that requires looking around.

Thank you for all your hard work!

### PR DESCRIPTION
**Pull request by pwnall**
This fixes #432 for me.

### CODE REVIEW

File: cmake/CXXFeatureCheck.cmake [hunk 1]
```diff
@@ -26,11 +26,27 @@ function(cxx_feature_check FILE)
...
-          ${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/${FILE}.cpp
-          CMAKE_FLAGS ${BENCHMARK_CXX_LINKER_FLAGS}
-          LINK_LIBRARIES ${BENCHMARK_CXX_LIBRARIES})
+  if(CMAKE_CROSSCOMPILING)
+    try_compile(COMPILE_${FEATURE}
+            ${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/${FILE}.cpp
+            CMAKE_FLAGS ${BENCHMARK_CXX_LINKER_FLAGS}
+            LINK_LIBRARIES ${BENCHMARK_CXX_LIBRARIES})
+    if(COMPILE_${FEATURE})
+      message("-- Setting HAVE_${VAR} to 0 may fix build failures")
```
Comment (dominichamon) [inline cmake/CXXFeatureCheck.cmake]: nit: maybe a warning?

"If you see build failures due to cross compilation, try setting HAVE_${VAR} to 0" ?

Comment (pwnall) [inline cmake/CXXFeatureCheck.cmake]: Done. Thank you very much for the feedback!

**Comment by dominichamon** [comment]
Thank you!

**Comment by pwnall** [comment]
@dominichamon Thank you very much for the quick review!

**Comment by apmanol** [comment]
Hello, I don't know if I should continue this issue or start a new one, but during cross compiling I have the following problem:
```
cmake -DCMAKE_TOOLCHAIN_FILE=/home/manap/Projects/xpu/nmsngv3/tools/share/cmake/toolchain-mips-wibasip.cmake -DBENCHMARK_ENABLE_TESTING=OFF .. 
-- The C compiler identification is GNU 4.8.1
-- The CXX compiler identification is GNU 4.8.1
-- Check for working C compiler: /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-gcc
-- Check for working C compiler: /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-g++
-- Check for working CXX compiler: /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
crosssss compiling
-- Found Git: /usr/bin/git (found version "2.9.5") 
-- git Version: v1.2.0-491360b8-dirty
-- Version: 1.2.0
-- Performing Test HAVE_CXX_FLAG_STD_CXX11
-- Performing Test HAVE_CXX_FLAG_STD_CXX11 - Success
-- Performing Test HAVE_CXX_FLAG_WALL
-- Performing Test HAVE_CXX_FLAG_WALL - Success
-- Performing Test HAVE_CXX_FLAG_WEXTRA
-- Performing Test HAVE_CXX_FLAG_WEXTRA - Success
-- Performing Test HAVE_CXX_FLAG_WSHADOW
-- Performing Test HAVE_CXX_FLAG_WSHADOW - Success
-- Performing Test HAVE_CXX_FLAG_WERROR
-- Performing Test HAVE_CXX_FLAG_WERROR - Success
-- Performing Test HAVE_CXX_FLAG_PEDANTIC
-- Performing Test HAVE_CXX_FLAG_PEDANTIC - Success
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS - Success
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32 - Failed
-- Performing Test HAVE_CXX_FLAG_WFLOAT_EQUAL
-- Performing Test HAVE_CXX_FLAG_WFLOAT_EQUAL - Success
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WZERO_AS_NULL_POINTER_CONSTANT
-- Performing Test HAVE_CXX_FLAG_WZERO_AS_NULL_POINTER_CONSTANT - Success
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WD654
-- Performing Test HAVE_CXX_FLAG_WD654 - Failed
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Failed
-- Performing Test HAVE_CXX_FLAG_COVERAGE
-- Performing Test HAVE_CXX_FLAG_COVERAGE - Success
-- Performing Test HAVE_STD_REGEX
CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_STD_REGEX to 0
Call Stack (most recent call first):
  CMakeLists.txt:180 (cxx_feature_check)


-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_POSIX_REGEX to 0
Call Stack (most recent call first):
  CMakeLists.txt:182 (cxx_feature_check)


-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK
CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_STEADY_CLOCK to 0
Call Stack (most recent call first):
  CMakeLists.txt:190 (cxx_feature_check)


-- Performing Test HAVE_STEADY_CLOCK -- success
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: /projects/Thirdparty/benchmark/builds_mips2
/benchmark/builds_mips2 (master *%)$ make VERBOSE=1 
/usr/bin/cmake -H/projects/Thirdparty/benchmark -B/projects/Thirdparty/benchmark/builds_mips2 --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /projects/Thirdparty/benchmark/builds_mips2/CMakeFiles /projects/Thirdparty/benchmark/builds_mips2/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/projects/Thirdparty/benchmark/builds_mips2'
make -f src/CMakeFiles/benchmark.dir/build.make src/CMakeFiles/benchmark.dir/depend
make[2]: Entering directory '/projects/Thirdparty/benchmark/builds_mips2'
cd /projects/Thirdparty/benchmark/builds_mips2 && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /projects/Thirdparty/benchmark /projects/Thirdparty/benchmark/src /projects/Thirdparty/benchmark/builds_mips2 /projects/Thirdparty/benchmark/builds_mips2/src /projects/Thirdparty/benchmark/builds_mips2/src/CMakeFiles/benchmark.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/projects/Thirdparty/benchmark/builds_mips2'
make -f src/CMakeFiles/benchmark.dir/build.make src/CMakeFiles/benchmark.dir/build
make[2]: Entering directory '/projects/Thirdparty/benchmark/builds_mips2'
[  6%] Building CXX object src/CMakeFiles/benchmark.dir/benchmark.cc.o
cd /projects/Thirdparty/benchmark/builds_mips2/src && /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-g++  -DHAVE_POSIX_REGEX -DHAVE_STD_REGEX -DHAVE_STEADY_CLOCK -I/projects/Thirdparty/benchmark/include -I/projects/Thirdparty/benchmark/src -I/projects/Thirdparty/benchmark/src/../include  -std=c++11  -Wall  -Wextra  -Wshadow  -pedantic  -pedantic-errors  -Wfloat-equal  -fstrict-aliasing  -Wzero-as-null-pointer-constant  -Wstrict-aliasing   -o CMakeFiles/benchmark.dir/benchmark.cc.o -c /projects/Thirdparty/benchmark/src/benchmark.cc
^Csrc/CMakeFiles/benchmark.dir/build.make:62: recipe for target 'src/CMakeFiles/benchmark.dir/benchmark.cc.o' failed

```

And the question is should  flags `-DHAVE_POSIX_REGEX -DHAVE_STD_REGEX`  should be enabled since it failed? And the second one, is how can choose which one from cmake option. The attempts with `-DRUN_HAVE_POSIX_REGEX=0`  failed.

**Pull request merged**